### PR TITLE
Move memory provider to core

### PR DESCRIPTION
**Pull request by nyalldawson**
This commit introduces basic support for non-library based data providers, and moves the memory provider from a library based provider to a core provider.

The memory provider is used so frequently throughout QGIS that it has become integral to QGIS functionality and must be available wherever the QGIS core library is used.

The changes here (while not exposed yet to Python) could potentially be built on to allow creation of pure Python data providers.

### CODE REVIEW

**Comment by nyalldawson** [comment]
@wonder-sk assigning to you as I know you've expressed interest in this in the past

File: src/core/CMakeLists.txt [hunk 1]
```diff
@@ -1000,6 +1009,8 @@ INCLUDE_DIRECTORIES(
   metadata
   pal
   processing
+  providers
```
Comment (m-kuhn) [inline src/core/CMakeLists.txt]: Are you intending to move more providers here?

File: src/core/providers/memory/qgsmemoryproviderutils.cpp [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      return QStringLiteral( "string" );
+  }
+  return QStringLiteral( "string" ); // no warnings
+}
+
+QgsVectorLayer *QgsMemoryProviderUtils::createMemoryLayer( const QString &name, const QgsFields &fields, QgsWkbTypes::Type geometryType, const QgsCoordinateReferenceSystem &crs )
+{
+  QString geomType = QgsWkbTypes::displayString( geometryType );
+  if ( geomType.isNull() )
+    geomType = "none";
```
Comment (m-kuhn) [inline src/core/providers/memory/qgsmemoryproviderutils.cpp]: QStringLiteral?

File: src/core/CMakeLists.txt (hunk 1 above)
Comment (nyalldawson) [inline src/core/CMakeLists.txt]: Yes, ogr and gdal would be the other candidates

Comment (m-kuhn) [inline src/core/CMakeLists.txt]: So apparently memory provider can be used for custom python providers (though I can't really follow how exactly), that makes sense - although I'm not completely sold yet to the idea of (ab)using memory provider for this.

What's the advantage of integrating gdal/ogr in core?

Comment (nyalldawson) [inline src/core/CMakeLists.txt]: No, the provider is not exposed through the bindings, so end use remains unchanged. What's changed here is the requirement that all providers are loaded via libraries. Now it's possible to register providers directly with function pointers. So this should be moderatly easy to expand and expose to the bindings, hence allowing pure python data providers.

Comment (m-kuhn) [inline src/core/CMakeLists.txt]: I see, sounds good.

So what's the reason for shipping memory (and gdal/ogr) in core? Don't get me wrong, I don't see any downsides, just wondering if you've got some incredibly cool added value with shipping in core which I don't see.

Comment (nyalldawson) [inline src/core/CMakeLists.txt]: There's a couple of reasons

My main reason was I needed the memory provider utils part. And then it seemed odd that we'd have utilities for working with certain providers in core when those providers themselves aren't in core.

Then it became an exercise in seeing what was missing from enabling plugin based providers...

File: src/core/providers/memory/qgsmemoryproviderutils.cpp (hunk 2 above)
Comment (nyalldawson) [inline src/core/providers/memory/qgsmemoryproviderutils.cpp]: Fixed

**Review by wonder-sk — APPROVED** [review]
The code looks great, the memory layer creation will be very useful!

Optional suggestions :-)
- It may be good to update docstrings of functions in QgsProviderRegistry (library(), providerLibrary(), function()) and QgsProviderMetadata (library()) to indicate that library will be null if it is internal provider.
- The code in QgsMemoryProviderUtils::createMemoryLayer() could be maybe simplified a little bit - to use a list for all key/value pairs (including CRS) and then joining those - to avoid need for the 'first' variable.

**Comment by nyalldawson** [comment]
@wonder-sk 

> It may be good to update docstrings of functions in QgsProviderRegistry (library(), providerLibrary(), function()) and QgsProviderMetadata (library()) to indicate that library will be null if it is internal provider.

Fixed

> The code in QgsMemoryProviderUtils::createMemoryLayer() could be maybe simplified a little bit - to use a list for all key/value pairs (including CRS) and then joining those - to avoid need for the 'first' variable.

Thanks, fixed

**Pull request merged**